packages/expo-updates/ios/EXUpdates/AppLoader/EXUpdatesCodeSigningConfiguration.swift

```diff
@@ -0,0 +1,138 @@
+  }
+  
+  @objc
+  public func createAcceptSignatureHeader() -> String {
+    return "sig, keyid=\"\(keyId)\", alg=\"\(algorithm.rawValue)\""
+  }
+  
+  @objc
+  public func verifySignatureHeaderInfo(signatureHeaderInfo: EXUpdatesSignatureHeaderInfo, signedData: Data) throws -> NSNumber {
+    // check that the key used to sign the response is the same as the key embedded in the configuration
+    // TODO(wschurman): this may change for child certificates and development certificates
+    if (signatureHeaderInfo.keyId != self.keyId) {
+      throw EXUpdatesCodeSigningConfigurationError.KeyIdMismatchError
+    }
+    
+    // note that a mismatched algorithm doesn't fail early. it still tries to verify the signature with the
+    // algorithm specified in the configuration
+    if (signatureHeaderInfo.algorithm != self.algorithm) {
+      NSLog("Key with alg=\(signatureHeaderInfo.algorithm) from signature does not match client configuration algorithm, continuing")
+    }
+    
+    guard let certificate = SecCertificateCreateWithData(nil, certificateDataDer as CFData) else {
+      throw EXUpdatesCodeSigningConfigurationError.CertificateParseError
+    }
+    guard let publicKey = SecCertificateCopyPublicKey(certificate) else {
+      throw EXUpdatesCodeSigningConfigurationError.CertificateParseError
+    }
+    
+    guard let signatureData = Data(base64Encoded: signatureHeaderInfo.signature) else {
+      throw EXUpdatesCodeSigningConfigurationError.SignatureEncodingError
+    }
+    
+    let isValid = try self.verifyRSASHA256SignedData(signedData: signedData, signatureData: signatureData, publicKey: publicKey)
+    return isValid ? NSNumber(booleanLiteral: true) : NSNumber(booleanLiteral: false)
+  }
+  
+  private func sha256(data : Data) -> Data {
+      var digest = Data(count: Int(CC_SHA256_DIGEST_LENGTH))
+      data.withUnsafeBytes { bytes in
+        digest.withUnsafeMutableBytes { mutableBytes in
+          _ = CC_SHA256(bytes.baseAddress, CC_LONG(data.count), mutableBytes.bindMemory(to: UInt8.self).baseAddress)
+        }
+      }
+      return digest
+  }
+  
+  private func verifyRSASHA256SignedData(signedData: Data, signatureData: Data, publicKey: SecKey) throws -> Bool {
+    let hashBytes = self.sha256(data: signedData)
+    var error: Unmanaged<CFError>?
+    return SecKeyVerifySignature(publicKey, .rsaSignatureDigestPKCS1v15SHA256, hashBytes as CFData, signatureData as CFData, &error)
+  }
+  
+  private static let beginPemBlock = "-----BEGIN CERTIFICATE-----"
+  private static let endPemBlock   = "-----END CERTIFICATE-----"
+  
+  private static func decodeToDER(pem pemData: Data) -> Data? {
+    if let pem = String(data: pemData, encoding: .ascii), pem.contains(beginPemBlock) {
+      let lines = pem.components(separatedBy: .newlines)
+      var base64buffer  = ""
+      var certLine = false
+      for line in lines {
+        if line == endPemBlock {
+          certLine = false
```

Comment:
Do we want to bail out here? I think DER needs certificates to be separated.

Reply:
This function is taken from the ASN1Decoder framework (it's private there). From what I can tell a cert in PEM format is just base 64 data surrounded by `-----BEGIN CERTIFICATE-----` and `-----END CERTIFICATE-----` so this function just discards anything outside of the two. You are correct that if there is a case where there can be multiple certificates in one PEM file this function would break (combine them all), and it does look like combining multiple certificates into a single PEM is valid: https://www.digicert.com/kb/ssl-support/pem-ssl-creation.htm.

So I'll return here, though we should also throw if there is more than one cert in the PEM since that's not in the spec. Now I'm curious what happens on android.